follower/consensus_follower.go

```diff
@@ -0,0 +1,150 @@
+package consensus_follower
+
+import (
+	"context"
+	"sync"
+
+	"github.com/onflow/flow-go/cmd"
+	access "github.com/onflow/flow-go/cmd/access"
+	"github.com/onflow/flow-go/consensus/hotstuff/notifications/pubsub"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// ConsensusFollower is a standalone module run by third parties which provides
+// a mechanism for observing the block chain. It maintains a set of subscribers
+// and delivers block proposals broadcasted by the consensus nodes to each one.
+type ConsensusFollower interface {
+	// Run starts the consensus follower.
+	Run(context.Context)
+	// AddOnBlockFinalizedConsumer adds a new block finalization subscriber.
+	AddOnBlockFinalizedConsumer(pubsub.OnBlockFinalizedConsumer)
+}
+
+// Config contains the configurable fields for a `ConsensusFollower`.
+type Config struct {
+	nodeID               flow.Identifier // the node ID of this node
+	upstreamAccessNodeID flow.Identifier // the node ID of the upstream access node
+	bindAddr             string          // address to bind on
```

Comment:
fyi - will also need the upstream staked AN networking public key. I will add that here later.

Reply:
Actually, from our offline discussion my understanding was that we would only need the staked AN address? Given that, the networking public keys are exchanged when the nodes connect (which happens in the libp2p library, but I'd assume there is some way to access it),  and then the node ID can be found via the DHT?

Reply:
so my bad - we do need the public key because when a node in libp2p connects to the other node, it uses a [multiaddress](https://docs.libp2p.io/concepts/addressing/). The multiaddress contains the libp2p node id e.g. `/ip4/7.7.7.7/tcp/4242/p2p/**QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N**`.

The very first thing a node does after connecting to the other node is a TLS [handshake to create an encrypted connection](https://github.com/libp2p/specs/blob/master/tls/tls.md#libp2p-public-key-extension).
For this to work, the node initiating the connection must have the public key of the peer. From this public key it can derive the libp2p node id and verify that the node id of the peer matches.

Reply:
@vishalchangrani okay interesting, so to check my understanding basically the libp2p node id is just a hash of the public key right?

Reply:
and is the Flow node ID different from the libp2p node id?

Reply:
> @vishalchangrani okay interesting, so to check my understanding basically the libp2p node id is just a hash of the public key right?

correct..it is just a hash of the public key.

Reply:
> and is the Flow node ID different from the libp2p node id? If it is the same, then we would _only_ need to provide the node id since upon connection to a remote peer, the peer will provide the public key for you to verify right? https://github.com/libp2p/specs/blob/master/tls/tls.md#:~:text=Extension-,In%20order%20to%20prove,the%20private%20host%20key,-The

The Flow node id is different from the libp2p node id. The Flow node id is also a hash of the public key but its a different hash AND it may not be a hash in the future and could be something else.

Reply:
Okay, so to summarize, we do need all three of the following:
* Staked node's Flow 
* Staked node's public key (to verify upon connection that it is who it says it is)
* Staked node IP address (to connect to it in the first place)

Is this correct?

Reply:
@smnzhu So this should help:
https://github.com/onflow/flow-go/blob/6c4685b73e99206b199bd3c5ccc7dd7ced7c8517/network/p2p/keyTranslator.go#L34-L100

https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md#peer-ids

i.e.  Flow's Network Public Keys and LibP2P's public key are a marshalling standard away from each other and inter-convertible. 

The libp2p PeerID is a hash of the libp2p public key. BUT when that public key is less than 42 bytes (which is the case for the ECC keys we use, at 32 bytes), that hash is the identity, after which the key is base58 encoded.

Hence libp2p can [tell you the key for the node you have the ID](https://github.com/libp2p/go-libp2p-core/blob/master/peer/peer.go#L91).

So if "˜=" means equal modulo format change:

Flow network key ˜= libp2p Public key ˜= libp2p PeerID

Besides that, Flow's (internal) NodeID is a SHA3-256 hash of the (Flow) staking public key, as you can see here:
https://github.com/onflow/flow-go/blob/64a592f3a099d2a313c23cdeedc3eba2d7f09538/cmd/bootstrap/cmd/keys.go#L69

This should probably be in a doc somewhere (@vishalchangrani  if you have suggestions as to where to put this?)

Reply:
It will be best to include this as a code comment [here](https://github.com/onflow/flow-go/blob/master/network/p2p/keyTranslator.go).